File: app/src/main/java/com/yeonsik/fitnessapp/data/MassUnit.java
package com.yeonsik.fitnessapp.data;

import java.util.Locale;

/** Supported user-facing mass units. Stored workout mass remains canonical kilograms. */
public enum MassUnit {
    KG("kg", "킬로그램"),
    LB("lb", "파운드");

    public static final double KILOGRAMS_PER_POUND = 0.45359237d;

    private final String id;
    private final String labelKo;

    MassUnit(String id, String labelKo) {
        this.id = id;
        this.labelKo = labelKo;
    }

    public String id() {
        return id;
    }

    public String symbol() {
        return id;
    }

    public String labelKo() {
        return labelKo;
    }

    public static MassUnit parse(String value) {
        if (value == null) {
            return null;
        }
        String normalized = value.trim().toLowerCase(Locale.ROOT);
        for (MassUnit unit : values()) {
            if (unit.id.equals(normalized)) {
                return unit;
            }
        }
        return null;
    }

    public static MassUnit orDefault(MassUnit unit) {
        return unit == null ? KG : unit;
    }

    public static double toKg(double value, MassUnit unit) {
        MassUnit effective = orDefault(unit);
        return effective == LB ? value * KILOGRAMS_PER_POUND : value;
    }

    public static double fromKg(double kilograms, MassUnit unit) {
        MassUnit effective = orDefault(unit);
        return effective == LB ? kilograms / KILOGRAMS_PER_POUND : kilograms;
    }

    public static double convert(double value, MassUnit from, MassUnit to) {
        return fromKg(toKg(value, from), to);
    }

    public double toKg(double value) {
        return toKg(value, this);
    }

    public double fromKg(double kilograms) {
        return fromKg(kilograms, this);
    }
}

File: shared/src/commonMain/kotlin/com/yeonsik/fitness/shared/feature/workout/model/MassUnit.kt
package com.yeonsik.fitness.shared.feature.workout.model

import kotlin.jvm.JvmField
import kotlin.jvm.JvmStatic

/** Supported user-facing mass units. Stored workout mass remains canonical kilograms. */
enum class MassUnit(
    private val persistedId: String,
    private val koreanLabel: String
) {
    KG("kg", "킬로그램"),
    LB("lb", "파운드");

    val id: String
        get() = persistedId

    fun id(): String = persistedId

    val labelKo: String
        get() = koreanLabel

    fun labelKo(): String = koreanLabel

    fun symbol(): String = persistedId

    companion object {
        @JvmField
        val KILOGRAMS_PER_POUND: Double = 0.45359237

        @JvmStatic
        fun parse(value: String?): MassUnit? {
            if (value == null) return null
            val normalized = value.trim().lowercase()
            return values().firstOrNull { it.persistedId == normalized }
        }

        @JvmStatic
        fun orDefault(unit: MassUnit?): MassUnit = unit ?: KG

        @JvmStatic
        fun toKg(value: Double, unit: MassUnit?): Double {
            val effective = orDefault(unit)
            return if (effective == LB) value * KILOGRAMS_PER_POUND else value
        }

        @JvmStatic
        fun fromKg(kilograms: Double, unit: MassUnit?): Double {
            val effective = orDefault(unit)
            return if (effective == LB) kilograms / KILOGRAMS_PER_POUND else kilograms
        }

        @JvmStatic
        fun convert(value: Double, from: MassUnit?, to: MassUnit?): Double =
            fromKg(toKg(value, from), to)
    }

    fun toKg(value: Double): Double = toKg(value, this)

    fun fromKg(kilograms: Double): Double = fromKg(kilograms, this)
}
